xc7/fasm2bels/iob_models.py

```diff
@@ -159,38 +158,42 @@ def append_ibuf_iostandard_params(
         top, site, bel, possible_iostandards, in_term=None
 ):
     """
-    Appends IOSTANDARD parameter to the bel. The parameter has to be explicitly
-    provided in the top.iostandard_defs dict. If the parameter from the dict
-    contradicts the one decoded from fasm, an error is printed.
+    Appends IOSTANDARD parameter to the bel. The parameter has to be decoded
+    from the EBLIF file. If the parameter from the EBLIF contradicts the one
+    decoded from fasm, an error is printed.
     """
 
-    # Check if we have external information about the IOSTANDARD
-    if site.site.name in top.iostandard_defs:
-        iostd_def = top.iostandard_defs[site.site.name]
-        iostandard = iostd_def["IOSTANDARD"]
+    # Check if we have IO settings information for the site read from EBLIF
+    iosettings = top.get_site_iosettings(site.site.name)
 
-        # Check if this is possible according to decoded fasm
-        is_valid = iostandard in possible_iostandards
-        if not is_valid:
-            eprint(
-                "IOSTANDARD setting provided for {} do not match"
-                "its counterpart decoded from the fasm".format(site.site.name)
-            )
+    # We don't. Use the default IOSTANDARD
+    if iosettings is None:
+        iosettings = {"IOSTANDARD": top.default_iostandard}
 
-            eprint("Requested:")
-            eprint(" {}".format(iostandard))
+    iostandard = iosettings.get("IOSTANDARD", None)
 
-            eprint("Candidates are:")
-            for i in possible_iostandards:
-                eprint(" {}".format(i.ljust(15)))
-            eprint("")
+    # Check if this is possible according to decoded fasm
+    is_valid = iostandard in possible_iostandards
+    if not is_valid:
+        eprint(
+            "IOSTANDARD setting provided for {} do not match "
+            "its counterpart decoded from the fasm".format(site.site.name)
+        )
 
-            # Demote NSTD-1 to warning
-            top.disable_drc("NSTD-1")
+        eprint("Requested:")
+        eprint(" {}".format(iostandard))
 
-        # Valid
-        else:
-            bel.parameters["IOSTANDARD"] = '"{}"'.format(iostandard)
+        eprint("Candidates are:")
+        for i in possible_iostandards:
+            eprint(" {}".format(i.ljust(15)))
+        eprint("")
+
+        # Demote NSTD-1 to warning
+        top.disable_drc("NSTD-1")
```

Comment:
Under what circumstances should the code ever reach this path? Why is this not a failure?

Reply:
Generally this happens when the IOSTANDARD and DRIVE parameters provided externally (via eblif or cmd line parameters) do not match possibilities that can be inferred from the fasm.

The condition will also be triggered when there is no external IOSTANDARD and DRIVE information provided.

I'm not sure whether this should be treated as an error. After all the design can be imported to Vivado for inspection even with some IOs do not have any IO standard assigned.

Reply:
In all cases, the IOSTANDARD needs to be supplied or explicit from the bitstream, yes?  In the event that an input from Vivado is supplied, it should have the right data?

Reply:
Still don't understand why this is not an error.

Reply:
This is not an error (for fasm2bels) because if it was then the lack of external IOSTANDARD data would cause the fasm2bels flow to fail. Very often one wants to import a design to Vivado just to look at placement and routing, not necessarily generate a bitstream. Therefore I think that fasm2bels should print a warning that it's missing IOSTANDARD data, but not fail.

From the Vivado bitstream generation it is an error, correct IOSTANDARD has to be specified. If the information is not available to fasm2bels (no EBLIF file of the design given), then one has to supply it externally to Vivado somehow (via XDC ?).

I agree that maybe disabling `NSTD-1` in case of missing IOSTANDARD data is not a good idea. Thoughts?

Now that you mentioned it: When decoding a Vivado bitstream with fasm2bels there is no EBLIF to take IOSTANDARD from. I'm not sure how to supply that data to fasm2bels in this case. Through the XDC file of the original design? Do you have any other ideas?

Reply:
But all bitstreams have an iostandard?  Therefore if the iostandard cannot be determined, that **is** a fatal error for fasm2bels

Reply:
But all bitstreams have an iostandard?  Therefore if the iostandard cannot be determined, that **is** a fatal error for fasm2bels

Reply:
> IOSTANDARD data would cause the fasm2bels flow to fail. Very often one wants to import a design to Vivado just to look at placement and routing

But how can you be confident in fasm2bels unless it generates the same bitstream?

Reply:
> Now that you mentioned it: When decoding a Vivado bitstream with fasm2bels there is no EBLIF to take IOSTANDARD from. I'm not sure how to supply that data to fasm2bels in this case. Through the XDC file of the original design? Do you have any other ideas?

This is a totally valid question, which we can file a feature request for and fix in a future version.